apps/extension/src/content-scripts/injected-penumbra-global.ts

```diff
@@ -5,54 +5,101 @@
- * simple api. The identifiers on this record should be unique and correspond to
- * an id in a manifest, and providers should provide a link to the manifest in
- * their record entry.
+ * simple API. The identifiers on this record should be unique, and correspond
+ * to a browser extension id. Providers should provide a link to their extension
+ * manifest in their record entry.
  *
- * The global is frozen to prevent mutation, but you should consider that the
+ * The global is frozen to discourage mutation, but you should consider that the
  * global and everything on it is only as trustable as the scripts running on
- * the page. Imports, includes, and packages your webapp depends on could all
- * mutate or preempt the global. User-agent injections like userscripts or other
- * content scripts could interfere or intercept connections.
+ * the page. Imports, requires, includes, script tags, packages your webapp
+ * depends on, userscripts, or other extensions' content scripts could all
+ * mutate or preempt this, and all have the power to interfere or intercept
+ * connections.
  */
 
 import { PenumbraInjection, PenumbraRequestFailure, PenumbraSymbol } from '@penumbra-zone/client';
-import { isPraxFailureMessageEvent, isPraxPortMessageEvent, PraxMessage } from './message-event';
+
+import {
+  isPraxFailureMessageEvent,
+  isPraxPortMessageEvent,
+  PraxMessage,
+  unwrapPraxMessage,
+} from './message-event';
 
 import { PraxConnection } from '../message/prax';
 
-const request = Promise.withResolvers<void>();
-
-// this is just withResolvers, plus a sync-queryable state attribute
-const connection = Object.assign(Promise.withResolvers<MessagePort>(), { state: false });
-connection.promise.then(
-  () => {
-    connection.state = true;
-    request.resolve();
-  },
-  () => {
-    connection.state = false;
-    request.reject();
-  },
-);
-
-// this resolves the connection promise when the isolated port script indicates
-const connectionListener = (msg: MessageEvent<unknown>) => {
-  if (isPraxPortMessageEvent(msg) && msg.origin === window.origin) {
-    // @ts-expect-error - ts can't understand the injected string
-    const praxPort: unknown = msg.data[PRAX];
-    if (praxPort instanceof MessagePort) connection.resolve(praxPort);
+type PromiseSettledResultStatus = PromiseSettledResult<unknown>['status'];
+
+class PraxInjection {
+  private static singleton?: PraxInjection = new PraxInjection();
+
+  public static get penumbra() {
+    return PraxInjection.singleton!.injection;
+  }
+
+  private manifestUrl = `${PRAX_ORIGIN}/manifest.json`;
+  private _request = Promise.withResolvers<void>();
+  private _connect = Promise.withResolvers<MessagePort>();
+  private _disconnect = Promise.withResolvers<void>();
+
+  private connectState?: PromiseSettledResultStatus;
+  private requestState?: PromiseSettledResultStatus;
+  private disconnectState?: PromiseSettledResultStatus;
```

Comment:
In general, it's quite difficult trying to crawl the lifecycle of these promises in the code. Is there a different paradigm or abstraction we could use to make this more maintainable? Maybe even consider a finite state machine pattern. Or maybe create helper classes of some kind.

Reply:
this is the helper class. previously all of this behavior was in a flat file.

we can return to that if you prefer.

Reply:
the `state` function was three lines long and essentially contained all of the synchronous state resolution logic from these three state inputs. it's expanded now in response to comments, but functions the same.

the `connectionFailure` getter contains logic to provide an asynchronous failure state from the promises whose state is tracked by these three state inputs.

was waiting to write documentation until after review, as writing documentation is significant effort and most of my suggested changes to this interface are not accepted.